izpack-compiler/src/main/java/com/izforge/izpack/compiler/packager/impl/PackagerBase.java

```diff
@@ -381,6 +384,41 @@ protected boolean packSeparateJars()
         return info != null && info.getWebDirURL() != null;
     }
 
+    private List<DynamicVariable> buildVariableList()
+    {
+        DirectedGraph<DynamicVariable> graph = new DirectedGraph<DynamicVariable>();
+        DynamicVariable rootVar = new DynamicVariableImpl("", "");
+
+        for (List<DynamicVariable> dynVariables : dynamicVariables.values())
+        {
+            for (DynamicVariable var : dynVariables)
+            {
+                graph.addEdge(rootVar, var);
+                for (String childName : var.getUnresolvedVariableNames())
+                {
+                    List<DynamicVariable> childVars = dynamicVariables.get(childName);
+                    if (childVars != null)
+                    {
+                        for (DynamicVariable childVar : childVars)
+                        {
+                            graph.addEdge(var, childVar);
+                        }
+                    }
+                }
+            }
+        }
+
+        Iterator<DynamicVariable> it = new BreadthFirstIterator<DynamicVariable>(graph, rootVar);
+        List<DynamicVariable> nodes = new ArrayList<DynamicVariable>();
+        while (it.hasNext()) {
+            nodes.add(it.next());
+        }
+        nodes.remove(rootVar);
```

Comment:
You're right. I had some ideal examples. The chosen traversal method is wrong.
Regarding the tree building - I agree also, using of the "root" node is not the right way.
So this implementation is crap and we can just reuse the implementation of the directed graph.

Regarding the "longest way" - it seems for more complex examples there may exist several "longest ways" which are all valid and it would be enough to get one of them (the above one has just one solution though). I'll try again to find another kind of tree traversal. A clean DFS search doesn't fulfill this also if the tree is not preordered somehow. If you find something useful for this problem let me know, please.